ocaml-lsp-server/src/hover_req.ml

```diff
@@ -316,111 +381,26 @@ let type_enclosing_hover ~(server : State.t Server.t) ~(doc : Document.t)
-    let at_decl_location (loc : Ocaml_parsing.Location.t) =
-      decl.ptype_loc.loc_start.pos_cnum <= loc.loc_start.pos_cnum
-      && decl.ptype_loc.loc_end.pos_cnum >= loc.loc_start.pos_cnum
-    in
-    let contains_decl (decls : Parsetree.type_declaration list) =
-      List.exists decls ~f:(fun (decl' : Parsetree.type_declaration) ->
-          String.equal decl.ptype_name.txt decl'.ptype_name.txt)
-    in
-    let signature_item (self : Ast_iterator.iterator)
-        (item : Parsetree.signature_item) =
-      match at_decl_location item.psig_loc with
-      | true -> (
-        match item.psig_desc with
-        | Psig_type (_, decls) when contains_decl decls ->
-          (* Don't add the type declaration itself. *) ()
-        | _ -> signature := item :: !signature)
-      | false -> Ast_iterator.default_iterator.signature_item self item
-    in
-    let structure_item (self : Ast_iterator.iterator)
-        (item : Parsetree.structure_item) =
-      match at_decl_location item.pstr_loc with
-      | true -> (
-        match item.pstr_desc with
-        | Pstr_type (_, decls) when contains_decl decls ->
-          (* Don't add the type definition itself. *) ()
-        | _ -> structure := item :: !structure)
-      | false -> Ast_iterator.default_iterator.structure_item self item
-    in
-    let iterator =
-      { Ast_iterator.default_iterator with signature_item; structure_item }
-    in
-    let () =
-      match ppx_parsetree with
-      | `Interface signature -> iterator.signature iterator signature
-      | `Implementation structure -> iterator.structure iterator structure
-    in
-    let expansion =
-      match (!signature, !structure) with
-      | [], [] -> None
-      | signature, [] ->
-        ignore (Format.flush_str_formatter ());
-        Pprintast.signature Format.str_formatter (List.rev signature);
-        Some (Format.flush_str_formatter ())
-      | [], structure ->
-        Some (Pprintast.string_of_structure (List.rev structure))
-      | _ :: _, _ :: _ ->
-        (* This should not be possible, unless a PPXs provides incorrect
-           position information that places items from a [sig end] into a
-           [struct end] or vice versa. *)
-        None
-    in
-    Option.map expansion ~f:(fun expansion ->
-        let range = Range.of_loc attr.attr_loc in
-        let contents =
-          format_ppx_expansion ~ppx:attr.attr_name.txt ~expansion
-        in
-        Hover.create ~contents ~range ())
-  | _ -> None
-
 let handle server { HoverParams.textDocument = { uri }; position; _ } mode =
   Fiber.of_thunk (fun () ->
       let state : State.t = Server.state server in
       let doc =
         let store = state.store in
         Document_store.get store uri
       in
+      let* ppxed_parsetree =
+        Document.Merlin.with_pipeline_exn
+          ~name:"expand-ppx"
+          (Document.merlin_exn doc)
+          (fun pipeline -> Mpipeline.ppx_parsetree pipeline)
+      in
```

Comment:
We should not retrived the ppxed_parstree if we are not going to use it, we should move it closer to the actual usage.